tests/System.Buffers.Primitives.Tests/MemoryTests.cs

```diff
@@ -235,6 +235,94 @@ public void AutoDispose()
             reservation.Dispose();
             Assert.Equal(true, owned.IsDisposed);
         }
+
+        [Fact]
+        public void PinAddReferenceReleaseTest()
+        {
+            var array = new byte[1024];
+            OwnedArray<byte> owned = array;
+            var memory = owned.Memory;
+            Assert.False(owned.HasOutstandingReferences);
+            var h = memory.Pin();
+            Assert.True(owned.HasOutstandingReferences);
+            h.Free();
+            Assert.False(owned.HasOutstandingReferences);
+        }
+
+        [Fact]
+        public void MemoryHandleFreeUninitialized()
+        {
+            var h = default(MemoryHandle);
+            h.Free();
+        }
+
+        [Fact]
+        public void MemoryHandleDoubleFree() 
+        {
+            var array = new byte[1024];
+            OwnedArray<byte> owned = array;
+            var memory = owned.Memory;
+            var h = memory.Pin();
+            Assert.True(owned.HasOutstandingReferences);
+            owned.AddReference();
+            Assert.True(owned.HasOutstandingReferences);
+            h.Free();
+            Assert.True(owned.HasOutstandingReferences);
+            h.Free();
+            Assert.True(owned.HasOutstandingReferences);
+            owned.Release();
+            Assert.False(owned.HasOutstandingReferences);
+        }
+
+
+        WeakReference LeakHandle()
+        {
+            // Creates an object that is both Pinned with a MemoryHandle,
+            // and has a weak reference.
+            var array = new byte[1024];
+            OwnedArray<byte> owned = array;
+            var memory = owned.Memory;
+            memory.Pin();
+            return new WeakReference(array);
+        }
+
+        WeakReference ReleaseHandle()
+        {
+            // Creates an object that is both Pinned with a MemoryHandle,
+            // and has a weak reference.
+            var array = new byte[1024];
+            OwnedArray<byte> owned = array;
+            var memory = owned.Memory;
+            var w = new WeakReference(array);
+            memory.Pin().Free();
+            return w;
+        }
+
+        void DoGC()
+        {
+            GC.Collect(2);
+            GC.WaitForPendingFinalizers();
+            GC.Collect(2);
+        }
+
+        [Fact]
+        void PinGCArrayTest1()
```

Comment:
Checking that something was not collected (PinGCArrayTest1) is reliable.

Checking that something got collected (PinGCArrayTest2) tends to have reliability issues. The object lifetime can be extended for variety of reasons: the codegen can decide to inline LeakHandle and extend the GC reporting lifetime for the locals; or conservative GC reporting can extend lifetime of arbitrary objects (conservative reporting its not used in CoreCLR today but it may change ... it is used in CoreRT in certain situations).